source/common/crypto/utility.h

```diff
@@ -25,50 +24,44 @@ struct VerificationOutput {
   std::string error_message_;
 };
 
-using PublicKeyPtr = bssl::UniquePtr<EVP_PKEY>;
+namespace Utility {
 
-class Utility {
-public:
-  /**
-   * Computes the SHA-256 digest of a buffer.
-   * @param buffer the buffer.
-   * @return a vector of bytes for the computed digest.
-   */
-  static std::vector<uint8_t> getSha256Digest(const Buffer::Instance& buffer);
-
-  /**
-   * Computes the SHA-256 HMAC for a given key and message.
-   * @param key the HMAC function key.
-   * @param message message data for the HMAC function.
-   * @return a vector of bytes for the computed HMAC.
-   */
-  static std::vector<uint8_t> getSha256Hmac(const std::vector<uint8_t>& key,
-                                            absl::string_view message);
+/**
+ * Computes the SHA-256 digest of a buffer.
+ * @param buffer the buffer.
+ * @return a vector of bytes for the computed digest.
+ */
+std::vector<uint8_t> getSha256Digest(const Buffer::Instance& buffer);
 
-  /**
-   * Verify cryptographic signatures.
-   * @param hash hash function(including SHA1, SHA224, SHA256, SHA384, SHA512)
-   * @param key pointer to public key
-   * @param signature signature
-   * @param text clear text
-   * @return If the result_ is true, the error_message_ is empty; otherwise,
-   * the error_message_ stores the error message
-   */
-  static const VerificationOutput verifySignature(absl::string_view hash, EVP_PKEY* key,
-                                                  const std::vector<uint8_t>& signature,
-                                                  const std::vector<uint8_t>& text);
+/**
+ * Computes the SHA-256 HMAC for a given key and message.
+ * @param key the HMAC function key.
+ * @param message message data for the HMAC function.
+ * @return a vector of bytes for the computed HMAC.
+ */
+std::vector<uint8_t> getSha256Hmac(const std::vector<uint8_t>& key, absl::string_view message);
 
-  /**
-   * Import public key.
-   * @param key key string
-   * @return pointer to public key
-   */
-  static PublicKeyPtr importPublicKey(const std::vector<uint8_t>& key);
+/**
+ * Verify cryptographic signatures.
+ * @param hash hash function(including SHA1, SHA224, SHA256, SHA384, SHA512)
+ * @param key pointer to EVP_PKEY public key
+ * @param signature signature
+ * @param text clear text
+ * @return If the result_ is true, the error_message_ is empty; otherwise,
+ * the error_message_ stores the error message
+ */
+const VerificationOutput verifySignature(absl::string_view hash, void* key,
+                                         const std::vector<uint8_t>& signature,
+                                         const std::vector<uint8_t>& text);
 
-private:
-  static const EVP_MD* getHashFunction(absl::string_view name);
-};
+/**
+ * Import public key.
+ * @param key key string
+ * @return pointer to EVP_PKEY public key
+ */
+void* importPublicKey(const std::vector<uint8_t>& key);
```

Comment:
Changing from UniquePtr to void* changed the semantics and leak the public key.

Reply:
I don't follow. Do you have an alternate suggestion for a signature? We can't use UniquePtr because it is BoringSSL-specific. I'm not a big fan of void*. How about std::unique_ptr?

Reply:
Nevermind, unique_ptr wont work since it won't free EVP_PKEY properly. I did add the EVP_PKEY_free() calls where appropriate in the existing code (and were caught by the tests). Using void* does put the responsibility on the caller of importPublicKey to free the EVP_PKEY. Add that to the doc? 

We have an openssl implementation of UniquePtr ... my only other idea would be to create a common.crypto.UniquePtr to wrap the EVP_PKEY but that seems messy and confusing.

Reply:
Will some of wrapper class that wraps `Envoy::CSmartPtr<EVP_PKEY, EVP_PKEY_free>` works here?

Reply:
Working on a wrapper class, thanks